internal/cluster/store/store_test.go

```diff
@@ -0,0 +1,317 @@
+	defer ln.Close()
+	if err := s.Open(); err != nil {
+		t.Fatalf("failed to open store: %s", err.Error())
+	}
+
+	if err := s.Bootstrap(&Node{
+		Id:         s.raftID,
+		Addr:       s.Addr(),
+		Partitions: map[uint32]NodePartition{},
+	}); err != nil {
+		t.Fatalf("failed to bootstrap single-node store: %s", err.Error())
+	}
+	if _, err := s.WaitForLeader(10 * time.Second); err != nil {
+		t.Fatalf("Error waiting for leader: %s", err)
+	}
+
+	testNodeId := "test-node"
+
+	s.WriteNodeChange(&proto.NodeChange{
+		NodeId: testNodeId,
+		Addr:   "",
+		State:  proto.NodeState_NODE_STATE_ERROR,
+		Role:   proto.Role_ROLE_TYPE_UNKNOWN,
+	})
+
+	// Snap the node and write to disk.
+	f := s.raft.Snapshot()
+	if f.Error() != nil {
+		t.Fatalf("failed to snapshot node: %s", f.Error())
+	}
+
+	snapDir := t.TempDir()
+	snapFile, err := os.Create(filepath.Join(snapDir, "snapshot"))
+	if err != nil {
+		t.Fatalf("failed to create snapshot file: %s", err.Error())
+	}
+	defer snapFile.Close()
+
+	fsm := NewFSM(s)
+	snapshot, err := fsm.Snapshot()
+	if err != nil {
+		t.Fatalf("failed to create snapshot from the store: %s", err)
+	}
+
+	sink := &mockSnapshotSink{snapFile}
+	if err = snapshot.Persist(sink); err != nil {
+		t.Fatalf("failed to persist snapshot to disk: %s", err.Error())
+	}
+
+	// Zero out the state
+	s.state = ClusterState{}
+
+	// Check restoration.
+	snapFile, err = os.Open(filepath.Join(snapDir, "snapshot"))
+	if err != nil {
+		t.Fatalf("failed to open snapshot file: %s", err.Error())
+	}
+	defer snapFile.Close()
+	if err := fsm.Restore(snapFile); err != nil {
+		t.Fatalf("failed to restore snapshot from disk: %s", err.Error())
+	}
+
+	// Ensure state is back in the correct state.
+	restoredNode, ok := s.state.Nodes[testNodeId]
+	if !ok {
+		t.Fatalf("failed to read test-node from restored snapshot")
+	}
+	if restoredNode.Id != "test-node" {
+		t.Fatalf("expected node Id to be %s was %s", testNodeId, restoredNode.Id)
+	}
+	if restoredNode.State != NodeStateError {
+		t.Fatalf("expected node Id to be %s was %s", testNodeId, restoredNode.Id)
+	}
+}
+
+// waitForLeaderID waits until the Store's LeaderID is set, or the timeout
+// expires. Because setting Leader ID requires Raft to set the cluster
+// configuration, it's not entirely deterministic when it will be set.
+func waitForLeaderID(s *Store, timeout time.Duration) (string, error) {
+	tck := time.NewTicker(100 * time.Millisecond)
+	defer tck.Stop()
+	tmr := time.NewTimer(timeout)
+	defer tmr.Stop()
+
+	for {
+		select {
+		case <-tck.C:
+			id, err := s.LeaderID()
+			if err != nil {
+				return "", err
+			}
+			if id != "" {
+				return id, nil
+			}
+		case <-tmr.C:
+			return "", fmt.Errorf("timeout expired")
+		}
+	}
+}
+
+type mockSnapshotSink struct {
+	*os.File
+}
+
+func (m *mockSnapshotSink) ID() string {
+	return "1"
+}
+
+func (m *mockSnapshotSink) Cancel() error {
+	return nil
+}
+
+func newMustTestStore(t *testing.T, c config.Cluster) (*Store, net.Listener) {
```

Comment:
I guess a `MuxTestStore` is meant?
```suggestion
func newMuxTestStore(t *testing.T, c config.Cluster) (*Store, net.Listener) {
```

Reply:
no `Must` is commonly used in tests and it indicates that we have to create the store successfully otherwise the functions panics/fails this is to prevent unnecessary error handling in test cases.

Reply:
Interesting pattern, never cross my way so far. Thanks for sharing.